src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Split.cs

```diff
@@ -0,0 +1,152 @@
+        public static SpanSplitEnumerator<T> Split<T>(this ReadOnlySpan<T> source, T separator)
+           where T : IEquatable<T> => new SpanSplitEnumerator<T>(source, separator);
+
+        /// <summary>
+        /// Returns a type that allows for enumeration of each element within a split span
+        /// using the provided separator span.
+        /// </summary>
+        /// <param name="source">The source span to be enumerated.</param>
+        /// <param name="separator">The separator span to be used to split the provided span.</param>
+        /// <returns>Returns a <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/>.</returns>
+        public static SpanSplitEnumerator<T> Split<T>(this ReadOnlySpan<T> source, ReadOnlySpan<T> separator)
+            where T : IEquatable<T> => new SpanSplitEnumerator<T>(source, separator, treatAsSingleSeparator: true);
+
+        /// <summary>
+        /// Returns a type that allows for enumeration of each element within a split span
+        /// using any of the provided elements.
+        /// </summary>
+        /// <param name="source">The source span to be enumerated.</param>
+        /// <param name="separators">The separators to be used to split the provided span.</param>
+        /// <returns>Returns a <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/>.</returns>
+        public static SpanSplitEnumerator<T> SplitAny<T>(this ReadOnlySpan<T> source, [UnscopedRef] params ReadOnlySpan<T> separators)
+            where T : IEquatable<T> => new SpanSplitEnumerator<T>(source, separators, treatAsSingleSeparator: false);
+
+        /// <summary>
+        /// Returns a type that allows for enumeration of each element within a split span
+        /// using the provided <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/>.
+        /// </summary>
+        /// <param name="source">The source span to be enumerated.</param>
+        /// <param name="separators">The <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/> to be used to split the provided span.</param>
+        /// <returns>Returns a <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/>.</returns>
+        public static SpanSplitEnumerator<T> SplitAny<T>(this ReadOnlySpan<T> source, SearchValues<T> separators)
+            where T : IEquatable<T> => new SpanSplitEnumerator<T>(source, separators);
+
+        /// <summary>
+        /// <see cref="System.MemoryExtensions.SpanSplitEnumerator{T}"/> allows for enumeration of each element within a <see cref="System.ReadOnlySpan{T}"/>
+        /// that has been split using a provided separator.
+        /// </summary>
+        public ref struct SpanSplitEnumerator<T> where T : IEquatable<T>
+        {
+            private enum SplitMode
+            {
+                None = 0,
+                SingleToken,
+                Sequence,
+                Any,
+                SearchValues
+            }
```

Comment:
This enum should be moved out of the generic type.

Reply:
Valid point, though not sure where to move it. Doesn't really seem too appropriate in `MemoryExtensions`. A separate non-generic `SpanSplitEnumerator`?

Reply:
Just as a private type in MemoryExtensions is fine.